File: labs-opensource/M15-code-execution-sandbox/starter/code_agent.py
"""
M15 Lab - Step 2: The Self-Debugging Code Agent
================================================
The model writes code; your executor runs it; errors are fed back.
Run: python code_agent.py
"""

import json

from openai import OpenAI
from sandbox_executor import SubprocessExecutor

client = OpenAI(base_url="http://localhost:11434/v1", api_key="ollama")

EXECUTE_PYTHON_TOOL = {
    "type": "function",
    "function": {
        "name": "execute_python",
        "description": (
            "Execute Python code in a sandbox and return stdout, stderr, and "
            "exit_code. Use print() for any output you want to see."
        ),
        "parameters": {
            "type": "object",
            "properties": {
                "code": {"type": "string", "description": "Complete, runnable Python code"},
                "timeout_seconds": {"type": "integer", "description": "Max runtime (default 10)"},
            },
            "required": ["code"],
        },
    },
}

SYSTEM_PROMPT = (
    "You are a data analysis assistant. When asked to compute something, "
    "write complete Python code and call execute_python. Always import every "
    "module you use. Print your final answer so it appears in stdout. If you "
    "receive an error, read it carefully, fix the code, and call "
    "execute_python again."
)


class CodeExecutionAgent:
    """ReAct-style agent that writes code, runs it, and self-debugs."""

    def __init__(self, executor=None, model: str = "mistral", max_retries: int = 3):
        # Executor is injected so DockerExecutor can replace it without changes here
        self.executor = executor or SubprocessExecutor()
        self.model = model
        self.max_retries = max_retries

    def run(self, user_request: str) -> str:
        """Run the agent, return the final answer string.

        TODO:
        messages = [{"role": "system", "content": SYSTEM_PROMPT},
                    {"role": "user", "content": user_request}]
        result = None
        For attempt in range(self.max_retries + 1):
          1. response = client.chat.completions.create(model=self.model,
                 messages=messages, tools=[EXECUTE_PYTHON_TOOL],
                 tool_choice="auto", temperature=0.1)
             ← low temp = more deterministic code
             (try/except → return f"API error: {e}")
          2. msg = response.choices[0].message
          3. If not msg.tool_calls: return msg.content or "(no output)"
             ← GOTCHA: some Ollama builds drop tool_choice silently;
               checking msg.tool_calls is the reliable signal
          4. messages.append(msg)   ← assistant message FIRST, then results
          5. For each tc in msg.tool_calls:
             - args = json.loads(tc.function.arguments)
             - result = self.executor.run(args["code"], args.get("timeout_seconds", 10))
             - print a short trace (e.g. first line of the code + exit_code)
             - messages.append({"role": "tool", "tool_call_id": tc.id,
                                "content": result.to_tool_content()})
          6. If attempt == self.max_retries and result and result.exit_code != 0:
               return f"Failed after {self.max_retries} attempts.\\n"
                      f"Last error:\\n{result.stderr}"
        After the loop: one final model call WITHOUT tools to summarize,
        return its content.
        """
        pass  # Remove this line when you add your code


# ── Smoke test (COMPLETE) ──
if __name__ == "__main__":
    agent = CodeExecutionAgent()

    print("TEST 1: computation the model can't do in its head")
    answer = agent.run("What is 3.7 to the power of 12? Show your work.")
    print(f"\nAgent answer: {answer[:300]}")

    print("\nTEST 2: a task that usually needs a debug round")
    answer = agent.run(
        "Compute the 25th Fibonacci number and the sum of the first 25 "
        "Fibonacci numbers. Print both."
    )
    print(f"\nAgent answer: {answer[:300]}")

File: labs-opensource/M15-code-execution-sandbox/starter/code_agent.js
/**
 * M15 Lab - Step 2: The Self-Debugging Code Agent (Node.js)
 * ==========================================================
 * Run: node code_agent.js
 */

import OpenAI from "openai";
import { SubprocessExecutor, toToolContent } from "./sandbox_executor.js";

const client = new OpenAI({ baseURL: "http://localhost:11434/v1", apiKey: "ollama" });

const EXECUTE_PYTHON_TOOL = {
  type: "function",
  function: {
    name: "execute_python",
    description:
      "Execute Python code in a sandbox and return stdout, stderr, and " +
      "exit_code. Use print() for any output you want to see.",
    parameters: {
      type: "object",
      properties: {
        code: { type: "string", description: "Complete, runnable Python code" },
        timeout_seconds: { type: "integer", description: "Max runtime (default 10)" },
      },
      required: ["code"],
    },
  },
};

const SYSTEM_PROMPT =
  "You are a data analysis assistant. When asked to compute something, " +
  "write complete Python code and call execute_python. Always import every " +
  "module you use. Print your final answer so it appears in stdout. If you " +
  "receive an error, read it carefully, fix the code, and call " +
  "execute_python again.";

class CodeExecutionAgent {
  constructor({ executor = null, model = "mistral", maxRetries = 3 } = {}) {
    // Executor is injected so a DockerExecutor can replace it without changes here
    this.executor = executor ?? new SubprocessExecutor();
    this.model = model;
    this.maxRetries = maxRetries;
  }

  /**
   * Run the agent, return the final answer string.
   *
   * TODO:
   * const messages = [{ role: "system", content: SYSTEM_PROMPT },
   *                   { role: "user", content: userRequest }];
   * let result = null;
   * For (attempt = 0; attempt <= this.maxRetries; attempt++):
   *   1. response = await client.chat.completions.create({ model: this.model,
   *        messages, tools: [EXECUTE_PYTHON_TOOL], tool_choice: "auto",
   *        temperature: 0.1 })   ← low temp = more deterministic code
   *      (try/catch → return `API error: ...`)
   *   2. const msg = response.choices[0].message;
   *   3. If (!msg.tool_calls?.length) return msg.content ?? "(no output)";
   *      ← some Ollama builds drop tool_choice silently; check tool_calls
   *   4. messages.push(msg);   ← assistant message FIRST, then results
   *   5. For each tc of msg.tool_calls:
   *      - const args = JSON.parse(tc.function.arguments);
   *      - result = await this.executor.run(args.code, args.timeout_seconds ?? 10);
   *      - log a short trace (first line of code + exitCode)
   *      - messages.push({ role: "tool", tool_call_id: tc.id,
   *                        content: toToolContent(result) });
   *   6. If (attempt === this.maxRetries && result && result.exitCode !== 0)
   *        return `Failed after ${this.maxRetries} attempts.\nLast error:\n${result.stderr}`;
   * After the loop: one final model call WITHOUT tools to summarize.
   */
  async run(userRequest) {
    // TODO: implement
  }
}

// ── Smoke test (COMPLETE) ──
const agent = new CodeExecutionAgent();

console.log("TEST 1: computation the model can't do in its head");
let answer = await agent.run("What is 3.7 to the power of 12? Show your work.");
console.log(`\nAgent answer: ${(answer ?? "").slice(0, 300)}`);

console.log("\nTEST 2: a task that usually needs a debug round");
answer = await agent.run(
  "Compute the 25th Fibonacci number and the sum of the first 25 Fibonacci numbers. Print both."
);
console.log(`\nAgent answer: ${(answer ?? "").slice(0, 300)}`);
